src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -284,9 +292,22 @@ instance
     FromMnemonic (mw ': '[]) purpose
   where
     fromMnemonic parts = do
-        -- FIXME Do better than 'show' here
-        m <- first (FromMnemonicError . show) (mkMnemonic @mw parts)
+        m <- first (FromMnemonicError . pretty) (mkMnemonic @mw parts)
         return $ Passphrase $ entropyToBytes $ mnemonicToEntropy m
+      where
+        pretty = \case
+            ErrMnemonicWords ErrWrongNumberOfWords{} ->
+                "Invalid number of words: at least "
+                <> show (natVal (Proxy :: Proxy mw))
+                <> " words are expected."
```

Comment:
Would it always say `at least 15` for `mnemonic_sentence` and `at least 9` for `mnemonic_second_factor` ?

Reply:
Indeed.

Reply:
Also for `mnemonic_second_factor` (where we allow 9 or 12) it will show somewhat faulty message for, say, 15 words I suppose...

Reply:
Maybe it could be a bit more comprehensive, like:
`Wrong number of words - N. Supported number of words are: mnemonic_sentence [15,18,21,24]; mnemonic_second_factor [9,12]`

Reply:
Hmm.. No :) 
The mnemonic length is actually encoded at the type-level here (that's what `mw` symbolizes) and, the error message is constructed based on this type. So, for _any_ representation `[9,12]` or `[15,18,21,24]` or `[12,18]` and so forth, the error will always report on the first (smallest) length.

Reply:
>Also for mnemonic_second_factor (where we allow 9 or 12) it will show somewhat faulty message for, say, 15 words I suppose...

I mean that when user provides 15 words for `mnemonic_second_factor` he will get `Invalid number of words: at least 9 words are expected`... well 15 is more than 9 so that's _somewhat misleading_

Reply:
Ah, got you! That requires slightly more work than it seems actually. Because of the way operations at the type-level work on list (at least, here), we do define instance in an inductive / recursive manner and, we "loose" (we actually consume them) some type-level information as we do it. I can try something for a few minutes, but let's scope it to 15 minutes ^^"